lib/pos_blockchain.go

```diff
@@ -134,8 +135,48 @@ func (bc *Blockchain) validateDeSoBlockPoS(desoBlock *MsgDeSoBlock) error {
 // that the timestamp is valid, that the version of the header is valid,
 // and other general integrity checks (such as not malformed).
 func (bc *Blockchain) validateBlockGeneral(desoBlock *MsgDeSoBlock) error {
-	// TODO: Implement me
-	return errors.New("IMPLEMENT ME")
+	// First make sure we have a non-nil header
+	if desoBlock.Header == nil {
+		return RuleErrorNilBlockHeader
+	}
+
+	// Timestamp validation
+	// What's the criteria? Is it just that it's not in the future?
+	// And that it's not too far in the past? Too far being older than the current best chain tip?
+	if desoBlock.Header.TstampNanoSecs < bc.GetBestChainTip().Header.TstampNanoSecs {
+		return RuleErrorPoSBlockTstampNanoSecsTooOld
+	}
+	if desoBlock.Header.TstampNanoSecs > uint64(time.Now().UnixNano()) {
+		return RuleErrorPoSBlockTstampNanoSecsInFuture
+	}
+
+	// Header validation
+	if desoBlock.Header.Version != HeaderVersion2 {
+		return RuleErrorInvalidPoSBlockHeaderVersion
+	}
+
+	// Malformed block checks
+	// Require header to have either vote or timeout QC
+	isTimeoutQCEmpty := desoBlock.Header.ValidatorsTimeoutAggregateQC.isEmpty()
+	isVoteQCEmpty := desoBlock.Header.ValidatorsVoteQC.isEmpty()
+	if isTimeoutQCEmpty && isVoteQCEmpty {
+		return RuleErrorNoTimeoutOrVoteQC
+	}
+
+	if !isTimeoutQCEmpty && !isVoteQCEmpty {
+		return RuleErrorBothTimeoutAndVoteQC
+	}
+
+	if !isVoteQCEmpty && len(desoBlock.Txns) == 0 {
+		return RuleErrorVoteQCWithoutTransactions
+	}
+
+	if !isTimeoutQCEmpty && len(desoBlock.Txns) != 0 {
+		return RuleErrorTimeoutQCWithTransactions
+	}
+
+	// TODO: What other checks do we need to do here?
```

Comment:
Some other validations that come to mind:
- Validate the txn merkle root
- Validate that the proposer's keys aren't nil/empty
- Validate that the proposer's random seed hash is properly formed
- Validate that the QCs are properly formed as well